File: 345/ReverseVowelsString.java
class ReverseVowelsString {
	String reverseVowels(String s) {
		char[] charArr;
		charArr = s.toCharArray();

		// collect all vowels in reverse
		StringBuilder vowelSB = new StringBuilder();
		for (char c: charArr) {
			if ("AEIOUaeiou".indexOf(c) >= 0)
				vowelSB.insert(0, c);
		}

		// fill in vowels appropriately
		StringBuilder ret = new StringBuilder();
		int i = 0;
		for (char c: charArr) {
			if ("AEIOUaeiou".indexOf(c) >= 0) {
				ret.append(vowelSB.charAt(i));
				i++;
			}
			else ret.append(c);
		}
		return ret.toString();
	}

	public static void main(String args[]) {
		ReverseVowelsString rvs = new ReverseVowelsString();
		
		String say = "Hello World!";
		System.out.println(say);
		System.out.println(rvs.reverseVowels(say));

		return;
	}
}

File: 345/reverse_vowels_string.py
class Solution(object):
    def reverseVowels(self, s):
        """
        :type s: str
        :rtype: str
        """
        VOWELS = "aeiouAEIOU"
        # extract all vowels
        vowels = "".join([v for v in s if v in VOWELS])
        
        # reverse extracted vowels
        vowels = vowels[::-1]
        
        # replace vowels
        ret = []
        i = 0
        for c in s:
            if c in VOWELS:
                ret.append(vowels[i])
                i += 1
            else:
                ret.append(c)
        return "".join(ret)